cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/engine/SQLEngine.java

```diff
@@ -0,0 +1,105 @@
+import io.cdap.cdap.etl.api.join.JoinDefinition;
+
+import java.util.concurrent.Future;
+
+/**
+ * A SQL Engine can be used to pushdown certain dataset operations.
+ *
+ * SQL Engines are implemented as plugins.
+ *
+ * Internally, the SQL Engine needs to handle retries and only surface {@link SQLEngineException} errors when there
+ * is a problem that cannot be recovered from and pipeline must be stopped.
+ *
+ * @param <KEY_OUT> The type for the Output Key when mapping a StructuredRecord
+ * @param <VALUE_OUT> The type for the Output Value when mapping a StructuredRecord
+ * @param <KEY_IN> The type for the Input Key when building a StructuredRecord
+ * @param <VALUE_IN> The type for the Input Value when building a StructuredRecord
+ */
+@Beta
+public interface SQLEngine<KEY_OUT, VALUE_OUT, KEY_IN, VALUE_IN> extends PipelineConfigurable,
+  SubmitterLifecycle<BatchContext>, KeyValueBiTransform<StructuredRecord, KEY_OUT, VALUE_OUT, KEY_IN, VALUE_IN> {
+  String PLUGIN_TYPE = "sqlengine";
+
+  /**
+   * Creates an Output Format Provided that can be used to push records into a SQL Engine.
+   *
+   * After created, this dataset will be considered "locked" until the output has been committed.
+   *
+   * @param datasetName The name of the dataset to use to store these records.
+   * @param datasetSchema the schema for this dataset.
+   * @return an {@link OutputFormatProvider} instance that can be used to write records to the SQL Engine into the
+   * specified dataset.
+   */
+  OutputFormatProvider getPushProvider(String datasetName, Schema datasetSchema) throws SQLEngineException;
+
+  /**
+   * Creates an InputFormatProvider that can be used to pull records from the specified dataset.
+   * @param datasetName the name of the dataset to pull records from.
+   * @param datasetSchema the schema for this dataset.
+   * @return AN {@link InputFormatProvider} instance that can be used to read records form the specified dataset in the
+   * SQL Engine.
+   */
+  InputFormatProvider getPullProvider(String datasetName, Schema datasetSchema) throws SQLEngineException;
+
+  /**
+   * Check if this dataset exists in the SQL Engine.
+   *
+   * This is a blocking call. if the process to write records into a dataset is ongoing, this method will block until
+   * the process completes. This ensures an accurate result for this operation.
+   *
+   * @param datasetName the dataset name.
+   * @return boolean specifying if this dataset exists in the remote engine.
+   */
+  boolean exists(String datasetName) throws SQLEngineException;
+
+  /**
+   * Check if the supplied Join Definition can be executed in this engine.
+   * @param definition the join definition to validate.
+   * @return boolean specifying if this join operation can be executed in the SQl Engine.
+   */
+  boolean canJoin(JoinDefinition definition);
+
+  /**
+   * Executes the join operation defined by the supplied join Definition and stores the result in the defined dataset.
```

Comment:
Do you plan to execute within this call? If yes, how would it work with chaining? I was expecting something like in spark where you build execution logic, but execution itself happens when you trigger write operation.

Reply:
Yes, invoking this method will trigger the BQ job to join multiple tables and condense results into another table.

The way we will handle chaining and async execution is by wrapping the SQL engine using the BatchSQLEngineAdapter and creating the concept of a SQL Engine Job that we can use to represent these operations that are running outside of spark. Users can use the SQL Engine Job API to check statuses and wait until the underlying tasks (Futures) have completed.

https://github.com/cdapio/cdap/blob/6cd828aa9fb930f5250afd0909b0c15d05bb988d/cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BatchSQLEngineAdapter.java#L53

Reply:
Got you. So, each step would be a separate SQL in SQLEngine. We are not going to merge multiple CDAP operations into a single SQL. Does it limit our potential performance gains? E.g. If we have join and then filter, we should be able to make it a single SQL (as soon as we support filters).

Reply:
this is a good question, I think we will probably end up with some sort of query representation instead of passing in a JoinDefinition.  There was a prototype that used Apache Calcite to do more sophisticated SQL manipulation (https://github.com/data-integrations/sql-pipelines) like combining multiple operations into a single query.  

Seeing that we want to only target joins in 6.5, I think it's ok to keep this for now, though it would be a good idea to mention in the javadoc that this API is expected to change in incompatible ways.

Reply:
Well, I'd just go with spark convention where after getting data in you have a  "SQLDataSet" that you can pass around and call transformations on, but any execution only happens when you need to get data out. This would allow SQL Engine to combine steps as needed.

Reply:
Basically, make `getPushProvider` return a `SQLWritableDataset` that extends `SQLDataSet` with `OutputFormatProvider`. `SQLDataSet` can be a marker interface, this would make user always ask SQLEngine to do the job or it can be methods of `SQLDataSet` itself. All other methods should take it instead of `dataSetName`. `join` would return another `SQLDataSet`. After you do pull you can ask for statistics either from every `SQLDataSet` directly or from `SQLEngine` passing each `SQLDataSet`. Or it can be done as a call back.

Reply:
@tivv That's a great suggestion. I will take a stab at this design in a later PR.

@albertshau I have left comments in my PR detailing that this method may cause compatibility issues in the future.